File: src/routes/pages.ts
import { Hono } from 'hono'
import cache from '../lib/cache.js'
import { sanitizeHtml } from '../lib/sanitize.js'
import M from '../modules/index.js'

const app = new Hono()

const VALID_TYPES = Object.keys(M)
const PAGE_TTL = 60 * 60 * 24 // 24 hours

app.get('/api/novel/:type/:id/pages/:num', async (c) => {
  const { type, id, num } = c.req.param()
  if (!VALID_TYPES.includes(type)) return c.json({ error: 'Invalid type' }, 400)
  const key = `novel:${type}:${id}:page:${num}`

  const cached = cache.get(key) as string | null
  if (cached !== null) return c.json({ html: cached })

  for (let i = 0; i < 3; i++) {
    try {
      const raw = await M[type].fetchPage(id, num)
      const html = sanitizeHtml(raw)
      cache.set(key, html, PAGE_TTL)
      return c.json({ html })
    } catch (e) {
      console.error(`fetchPage ${type}/${id}/${num} attempt ${i + 1} failed:`, e)
      if (i < 2) await new Promise(r => setTimeout(r, 500 * (i + 1)))
    }
  }
  return c.json({ error: 'Failed to fetch page' }, 502)
})

app.patch('/api/novel/:type/:id/pages/:num', async (c) => {
  const { type, id, num } = c.req.param()
  if (!VALID_TYPES.includes(type)) return c.json({ error: 'Invalid type' }, 400)
  const key = `novel:${type}:${id}:page:${num}`

  for (let i = 0; i < 3; i++) {
    try {
      const raw = await M[type].fetchPage(id, num)
      const html = sanitizeHtml(raw)
      cache.set(key, html, PAGE_TTL)
      return c.json({ html })
    } catch (e) {
      console.error(`fetchPage PATCH ${type}/${id}/${num} attempt ${i + 1} failed:`, e)
      if (i < 2) await new Promise(r => setTimeout(r, 500 * (i + 1)))
    }
  }
  return c.json({ error: 'Failed to fetch page' }, 502)
})

export default app

File: src/routes/pages.nim
import std/[asyncdispatch, json, options, logging, httpcore]
import ../error, ../state, ../cache, ../modules/module_type, ../sanitize

const PageTtl* = 60 * 60 * 24  # 24 hours

proc withRetryOpt(label: string, body: proc(): Future[Option[string]]): Future[Option[string]] {.async.} =
  for i in 0..2:
    try:
      return await body()
    except AppError:
      raise
    except CatchableError as e:
      error label & " attempt " & $(i + 1) & " failed: " & e.msg
      if i < 2:
        await sleepAsync(500 * (i + 1))
  raise newAppError(Upstream, "Failed after 3 retries: " & label)

proc fetchAndCache(state: AppState, module: ModuleType, id: string, num: string, key: string): Future[JsonNode] {.async.} =
  let label = "fetchPage " & id & "/" & num
  let raw = await withRetryOpt(label, proc(): Future[Option[string]] {.async.} =
    return await module.fetchPage(id, num)
  )
  let html = clean(raw.get(""))
  state.cache.put(key, %html, PageTtl)
  return %*{"html": html}

proc handleGetPage*(state: AppState, typeStr: string, id: string, num: string): Future[(HttpCode, string)] {.async.} =
  try:
    let module = resolve(typeStr)
    let key = "novel:" & typeStr & ":" & id & ":page:" & num
    let cached = state.cache.get(key)
    if cached.isSome:
      return (Http200, $(%*{"html": cached.get}))
    let res = await fetchAndCache(state, module, id, num, key)
    return (Http200, $res)
  except AppError as e:
    return (HttpCode(e.kind.statusCode), $e.toJson)
  except CatchableError:
    return (Http502, $(%*{"error": "Failed to fetch page"}))

proc handlePatchPage*(state: AppState, typeStr: string, id: string, num: string): Future[(HttpCode, string)] {.async.} =
  try:
    let module = resolve(typeStr)
    let key = "novel:" & typeStr & ":" & id & ":page:" & num
    let res = await fetchAndCache(state, module, id, num, key)
    return (Http200, $res)
  except AppError as e:
    return (HttpCode(e.kind.statusCode), $e.toJson)
  except CatchableError:
    return (Http502, $(%*{"error": "Failed to fetch page"}))
